src/libpmemfile-posix/block_array.c

```diff
@@ -0,0 +1,273 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "ctree.h"
+#include "layout.h"
+#include "inode.h"
+#include "internal.h"
+#include "out.h"
+
+static void
+update_first_block_info(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	if (binfo->arr != NULL) {
+		/*
+		 * If the block_info is not null, it means it was kept
+		 * up-to-date by the allocating routines.
+		 */
+		return;
+	}
+
+	/*
+	 * If binfo was not used before, it must be initialized.
+	 */
+
+	/*
+	 * Find the block_array containing the next free block metadata
+	 * slot. This is either the block_array stored right in the
+	 * inode, ...
+	 */
+	binfo->arr = &vinode->inode->file_data.blocks;
+	/*
+	 * ... or if there is more than one block_array, it is
+	 * the one linked to it with the next field.
+	 */
+	if (!TOID_IS_NULL(binfo->arr->next))
+		binfo->arr = D_RW(binfo->arr->next);
+
+	binfo->idx = 0;
+
+	/* Find the first free entry in the block array. */
+	while (binfo->idx < binfo->arr->length &&
+	    binfo->arr->blocks[binfo->idx].size != 0)
+		binfo->idx++;
+}
+
+static bool
+has_free_block_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->idx < binfo->arr->length;
+}
+
+static void
+allocate_new_block_array(struct pmemfile_vinode *vinode)
+{
+	TOID(struct pmemfile_block_array) new =
+			TX_ZALLOC(struct pmemfile_block_array, FILE_PAGE_SIZE);
+	D_RW(new)->length = (uint32_t)
+			((page_rounddown(pmemobj_alloc_usable_size(new.oid)) -
+			sizeof(struct pmemfile_block_array)) /
+			sizeof(struct pmemfile_block));
+
+	D_RW(new)->next = vinode->inode->file_data.blocks.next;
+	TX_SET_DIRECT(&vinode->inode->file_data.blocks, next, new);
+	vinode->first_free_block.arr = D_RW(new);
+	vinode->first_free_block.idx = 0;
+}
+
+/*
+ * is_zeroed -- check if given memory range is all zero
+ */
+static inline bool
+is_zeroed(const void *addr, size_t len)
+{
+	/* XXX optimize */
+	const char *a = (const char *)addr;
+	while (len-- > 0)
+		if (*a++)
+			return false;
+	return true;
+}
+
+static struct pmemfile_block *
+acquire_new_entry(struct pmemfile_vinode *vinode)
+{
+	ASSERT(has_free_block_entry(vinode));
+
+	struct block_info *binfo = &vinode->first_free_block;
+	struct pmemfile_block *block = binfo->arr->blocks + binfo->idx++;
+
+	ASSERT(is_zeroed(block, sizeof(*block)));
+
+	/* XXX, snapshot separated to let pmemobj use small object cache  */
+	pmemobj_tx_add_range_direct(block, 32);
```

Comment:
I know. I can't remove it yet, because with current pmemobj version pmemfile tests fail, and I can't fix them without allocation classes.